### DISCUSSION TITLE
get_tags_from_EH API $uri

### DISCUSSION THREAD

**Discussion by uap4** [Difegue/LANraragi#484] [General]
I tried LANraragi. It looks like my ISP is blocking E-Hentai with subdomains (PR_CONNECT_RESET_ERROR browser error) but not ExHentai. Then even with $enablepanda, because https://github.com/Difegue/LANraragi/blob/f22de134ad7351235fa3ba254604793684d583dd/lib/LANraragi/Plugin/Metadata/EHentai.pm#L234, request failed. I changed to "ex", but then there were redirects to the forum. Then I changed $ua  in https://github.com/Difegue/LANraragi/blob/f22de134ad7351235fa3ba254604793684d583dd/lib/LANraragi/Plugin/Metadata/EHentai.pm#L236 so that it was like in 
https://github.com/Difegue/LANraragi/blob/f22de134ad7351235fa3ba254604793684d583dd/lib/LANraragi/Plugin/Metadata/EHentai.pm#L83. I don't really understand how, but it works. 

It is likely that there will be another problem that when requesting https://e-hentai.org/api.php, redirection to CloudFlare (I've only tested this in the Firefox browser.). But the EHwiki contains an address on the third-level domain https://api.e-hentai.org/api.php, and when accessing it through a browser, no redirection occurs. Therefore, it is better to use https://api.e-hentai.org/api.php address?

For myself, I will probably copy the plugin and write it in my own way.

**Comment by Difegue** [comment]
As you've seen, the EH API is not available under the `ex` context.  
The cloudflare block happened a few weeks ago already, but had been removed later on: I suspect EH is now applying CF firewalls whenever it encounters too much traffic at a given time.  

I'd seen that the CF block was affecting the API URL we used, but to my knowledge the `api.e-hentai.org/api.php` TLD did not exist at the time! If it's been explicitly added as an alwayys-unblocked endpoint now, the LRR plugin should use it as a default.

**Reply by uap4** [reply]
Well I do not know.

>As you've seen, the EH API is not available under the ex context.

It works in LANraragi with my modifications. And from Firefox I can open https://exhentai.org/api.php, receive {"error": "Empty JSON Request"}, in F12 press edit and send again, select POST, insert json, received correct  {"gmetadata". 

>TLD did not exist at the time! If it's been explicitly added as an alwayys-unblocked endpoint now, the LRR plugin should use it as a default.

https://ehwiki.org/action/history/API  
Revision as of 16:42, 15 January 2016     -  API URL: http://g.e-hentai.org/api.php
Revision as of 13:27, 25 January 2017     - API URL: https://e-hentai.org/api.php  
Revision as of 21:20, 25 February 2017 - API URL: https://api.e-hentai.org/api.php

While it somehow works with my changes - it's okay for me. I hope that it will not break.

**Reply by Difegue** [reply]
Hmm, you're right that the url hasn't changed in the wiki in a while -- I think I tested both forms during the first CF block and both were blocked, but maybe that's been fixed now. 👍 

Similarly, I never got the api to work on ex before, but I guess it does work now if you give it the proper cookies! Feel free to push your changes to the plugin as a PR if you have the time, I'll gladly integrate them. 😁 

(If not I'll try to get this stuff improved later today)

**Reply by uap4** [reply]
>Feel free to push your changes to the plugin as a PR

Sketched out adding a custom API URI field. I don't know much about perl and github, so it's not good for PR. But for example, you can see the comparison, probably.  https://github.com/Difegue/LANraragi/compare/dev...uap4:patch-1

Also in the description `g.` not relevant for a long time. 
https://github.com/Difegue/LANraragi/blob/c287836f5a4fe61f8f1f76ef9886a98baf4c2942/lib/LANraragi/Plugin/Metadata/EHentai.pm#L29

**Reply by Difegue** [reply]
That'd work and actually looks pretty good code-wise, although I'm not sure the API field changes often enough to warrant making it a setting. 😛 

It's true that `g.` hasn't been a thing forever, I'll think about editing the description later.

**Comment by Difegue** [comment]
Improved in 11738260f7f62299dd92d561c90aa908d02cdc99.
